File: c_ds/train/basics/permutations.cpp
#include <iostream>
#include <vector>

using namespace std;

template <typename T, typename K>
T& operator<<(T& stream, vector<K>values){
    stream << "[";
    for (auto val: values){
        stream << val << ",";
    }
    stream << "]";
    return stream;
}

class Solution {
public:
    void dfs(vector<int>& nums, vector<int>& currPermute, vector<vector<int>>& result){
        if (nums.size() > 0) {
            for (auto num: nums){
                vector<int> numsLess1;
                for ( auto val : nums){
                    if (val != num) numsLess1.push_back(val);
                }
                currPermute.push_back(num);
                dfs(numsLess1, currPermute, result);
                currPermute.erase(--currPermute.end());
            }
        }else{
            result.push_back(currPermute);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        
        for (auto num: nums){
            vector<int> currPerm;
            vector<int> numsLess1;
            currPerm.push_back(num);
            for ( auto val : nums){
                if (val != num) numsLess1.push_back(val);
            }
            dfs(numsLess1, currPerm, result);
        }
        return result;
    }
};

void test(){
    vector<int> nums {1,2,3};
    Solution s;
    auto permutations = s.permute(nums);
    cout << "\t result:" << permutations << endl;

}

int main(){
    test();
    return 0;
}

File: c_ds/train/basics/permutations.py
class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:
        def dfs(nums:list[int], permute_res:list[int], result):
            print(f"\t permute_res:{permute_res} - sub_list:{nums}")
            if len(nums) > 0:
                for num in nums:
                    new_nums = nums.copy()
                    new_nums.remove(num)
                    permute_res.append(num)
                    dfs(new_nums, permute_res, result)
                    permute_res.remove(num)
            else:
                print(f"\t\t -- finally adding:{permute_res}")
                # permute_res_copy = permute_res.copy()
                result.append(permute_res[:])

        result = []
        for num in nums:
            permute_res = []
            new_nums = nums.copy()
            new_nums.remove(num)
            permute_res.append(num)
            dfs(new_nums, permute_res, result)
        return result
        

def test():
    nums = [1,2,3]
    s = Solution()
    longest_str = s.permute(nums)
    print(f"\t result:{longest_str}")

if __name__ == "__main__":
    test()